app/celery/process_delivery_status_result_tasks.py

```diff
@@ -265,3 +276,266 @@ def sms_status_update(
+        'Final %s logic | reference: %s | notification_id: %s | status: %s | status_reason: %s | cost_in_millicents: %s',
+        sms_status.provider,
+        sms_status.reference,
+        notification.id,
+        notification.status,
+        notification.status_reason,
+        notification.cost_in_millicents,
+    )
+
+    # Our clients are not prepared to deal with pinpoint payloads
+    if not _get_include_payload_status(notification):
+        sms_status.payload = None
+
+    try:
+        check_and_queue_callback_task(notification, sms_status.payload)
+    except Exception:
+        current_app.logger.exception('Failed check_and_queue_callback_task for notification: %s', notification.id)
+        statsd_client.incr(f'clients.sms.{sms_status.provider}.status_update.error')
+    else:
+        try:
+            check_and_queue_va_profile_notification_status_callback(notification)
+            statsd_client.incr(f'clients.sms.{sms_status.provider}.status_update.success')
+        except Exception:
+            current_app.logger.exception(
+                'Failed check_and_queue_va_profile_notification_status_callback for notification: %s',
+                notification.id,
+            )
+            statsd_client.incr(f'clients.sms.{sms_status.provider}.status_update.error')
+
+
+def sms_attempt_retry(
+    sms_status: SmsStatusRecord,
+    event_timestamp: str | None = None,
+    event_in_seconds: int = 300,  # Don't retry _get_notification by default
+):
+    """Attempt retry sending notification.
+
+    Retry notification if within permissible limits and update notification.
+    If retry limit or retry window exceeded call sms_status_update with STATUS_REASON_UNDELIVERABLE
+
+    Args:
+        sms_status (SmsStatusRecord): The status record update
+        event_timestamp (str | None, optional): Timestamp the Pinpoint event came in. Defaults to None.
+        event_in_seconds (int, optional): How long since initial event. Defaults to 300
+
+    Raises:
+        NonRetryableException: Unable to update SMS retry count or update the notification
+    """
+
+    # avoid circular import
+    from app.notifications.process_notifications import send_notification_to_queue_delayed
+
+    notification = _get_notification(sms_status.reference, sms_status.provider, event_timestamp, event_in_seconds)
+
+    current_app.logger.info(
+        'Entering %s retryable failure logic to process event | reference: %s | notification_id: %s | current_status: %s | current_status_reason: %s | event_sms_status: %s | event_sms_status_reason: %s',
+        sms_status.provider,
+        sms_status.reference,
+        notification.id,
+        notification.status,
+        notification.status_reason,
+        sms_status.status,
+        sms_status.status_reason,
+    )
+
+    notification_retry_id = f'notification-carrier-sms-retry-count-{notification.id}'
+
+    try:
+        retry_count_redis_ttl = int(CARRIER_SMS_MAX_RETRY_WINDOW.total_seconds())
```

Comment:
This statement should not raise an exception.  Please move it to before the "try".